[Sia Zhong Qing] iP

starsia opened the pull request:
## Klara 
> “Your mind is for having ideas, not holding them.” – _David Allen_ ([source](https://dansilvestre.com/productivity-quotes))

**Klara** frees your mind of having to remember things you need to do. It's,
- text-based ⌨️ 
- easy to learn 👍 
- ~~FAST~~ _SUPER FAST_ to use ⏩ 
- All you need to do is,

download it from [here](https://nus-cs2103-ay2324s2.github.io/website/schedule/week4/project.html).
1. double-click it. 🖱️ 
2. add your tasks. 🗒️ 
3. let it manage your tasks for you 😉
4. And it is FREE! 😮 

Features:
- [x] Managing tasks
- [x] Managing deadlines
- [x] Managing events
- [x] Reminders (coming soon)

If you are a Java programmer, you can use it to practice Java too. Here's the main method:

``` java
public class Main {
    public static void main(String[] args) {
        Application.launch(Duke.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

Commit:
Add Rename, Greet, Exit

Commit:
Add Echo

Commit:
Add "List" and "Add" commands

Commit:
Add Mark as Done

Commit:
Add ToDos, Events, Deadlines

Commit:
Add Automated Text UI Testing

Commit:
Revert "Add Automated Text UI Testing"

This reverts commit 212b1c26ade87f934ab6b64b0a8680605b0efde3.

Commit:
Add Automated Text UI Testing

Commit:
Add error handling

Commit:
Add Delete

Commit:
Add Enums

Commit:
Add Save

Dethada commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,205 @@
+        // Load tasks from file
+        // Solution below inspired by https://stackoverflow.com/questions/16111496/java-how-can-i-write-my-arraylist-to-a-file-and-read-load-that-file-to-the
+        try (ObjectInputStream ois = new ObjectInputStream(new FileInputStream("./data/duke.txt"))) {
+            list = (ArrayList<Task>) ois.readObject();
```
perhaps a more specific variable name like `taskList` would be better

Dethada commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,205 @@
+                new File("./data").mkdirs();
+                // then make file
+                // Solution below inspired by https://www.w3schools.com/java/java_files_create.asp
+                File myObj = new File("./data/duke.txt");
```
Similar to above, consider `file` or `dataFile`?

Dethada commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,34 @@
+    public String toString() {
+        if (isDone) {
+            return "[X] " + description;
+        }
+        else {
```
Similar to above, formatting of conditionals should be
```
if (condition) {
    statements;
} else {
    statements;
}
```

starsia replied:
Great spot! Thanks!

Dethada commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,205 @@
+        for (int i = 0; i < list.size(); i++) {
+            if (list.get(i).isDone) {
+                System.out.println(Integer.toString(i + 1) + "." + list.get(i));
+            }
+            else {
```
Following our coding standard, conditionals should be of the following form

```
if (condition) {
    statements;
} else {
    statements;
}
```

Dethada commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,205 @@
+        printBreak();
+
+        Scanner sc = new Scanner(System.in);
+        String word = sc.nextLine();
```
I think `word` is not the best variable name here, since we are reading in a line which can contain multiple words, consider `userInput` as a possible alternative

Dethada commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,205 @@
+                word = sc.nextLine();
+            // implementation of mark
+            } else if (word.startsWith("mark")) {
+                String[] split = word.split(" ");
```
Since `split` is a array of strings (words), consider renaming it to `splitWords`, there are many other occurrences of this below too.

starsia replied:
You're right. I've fixed that 🙇‍♂️

Dethada commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,205 @@
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.ObjectInputStream;
+import java.io.ObjectOutputStream;
+import java.util.ArrayList;
+import java.util.Scanner;
```
I like your organization of import statements here

Dethada reviewed (commented):
Coding standard looks good overall, other than small issues with variable naming and formatting of conditionals.

Commit:
Add Dates and Times

Commit:
Apply OOP to Commands: bye, mark, unmark, todo, delete

Commit:
Apply OOP to Commands: "event", "deadline"
Add Parser class
Fix TaskList saving upon "bye" command

Commit:
Add Use More OOP

Commit:
Organize into Packages
Opted to not branch and then merge this increment as it is a single step

Commit:
Correct Organize into Packages commit
Now, all classes are put into one package: duke

Commit:
Merge branch 'add-gradle-support'

* add-gradle-support:
  Bump gradle and lib version
  Add Gradle support
  docs/README.md: Tweak document template

Commit:
Use Gradle
Fix typo in output

Commit:
Add JUnit Tests

Commit:
Create JAR file (not committed)

tanqinyong commented on src/main/java/seedu/duke/Storage.java:
```diff
@@ -0,0 +1,53 @@
+        this.filepath = filepath;
+    }
+
+    public ArrayList<Task> load() throws DukeException {
```
Perhaps a better name would be "loadFile", which may help differentiate the method from similar future methods!

tanqinyong commented on src/main/java/seedu/duke/DukeException.java:
I like that you created your own exceptions!

tanqinyong commented on src/main/java/seedu/duke/Duke.java:
I like how neat and clean your Duke class is!

tanqinyong commented on src/main/java/seedu/duke/Storage.java:
```diff
@@ -0,0 +1,53 @@
+        return list;
+    }
+
+    public void saveList(ArrayList<Task> list) {
```
This is a good name!

tanqinyong reviewed (commented):
Overall, I find your code to be pretty complete and well-designed! It was easy to read and well-structured.

Commit:
Add JavaDoc Comments

Commit:
Follow Coding Standard: categorise import statement by package

Commit:
Attempt to fix Find

Commit:
Merge branch 'branch-A-JavaDoc'

* branch-A-JavaDoc:
  Add JavaDoc Comments

Commit:
In-progress fix for find command

Commit:
Fix coding standards from Git PRs

Commit:
Fix Checkstyle errors

Commit:
Fix bug NoClassDefFound Error

Commit:
Implement ChatBot Window layout (JavaFX Part 2)

Commit:
Add Functional GUI without tailored responses (JavaFX Part 3)

Commit:
Revert "Add Functional GUI without tailored responses (JavaFX Part 3)"

This reverts commit 186ad49400caf98b020eddcc4aa836a9a713dcef.

Commit:
Implement Java FX Part 3 Iteration 1

Commit:
Implement Java FX Part 3 Iteration 2

Commit:
Implement Java FX Part 3 Iteration 3 (Functional GUI)

Commit:
Add DialogBox.fxml and MainWindow.fxml

Commit:
Implement Java FX Part 4 (without Duke function)

Commit:
Convert Todo, Opening and Bye command to return type String

Commit:
Convert Mark, Unmark and Deadline command to return type String

Commit:
Convert Event, Delete to return type String
Add Duke responses to Klara

Commit:
Fix bugs for edge input cases

Commit:
Add opening message in GUI

Commit:
Code styling

Commit:
Merge branch 'master' of https://github.com/starsia/ip into branch-Level-9

* 'master' of https://github.com/starsia/ip:
  Code styling
  Add opening message in GUI
  Fix bugs for edge input cases
  Convert Event, Delete to return type String Add Duke responses to Klara
  Convert Mark, Unmark and Deadline command to return type String
  Convert Todo, Opening and Bye command to return type String
  Implement Java FX Part 4 (without Duke function)
  Add DialogBox.fxml and MainWindow.fxml
  Implement Java FX Part 3 Iteration 3 (Functional GUI)
  Implement Java FX Part 3 Iteration 2
  Implement Java FX Part 3 Iteration 1
  Revert "Add Functional GUI without tailored responses (JavaFX Part 3)"
  Add Functional GUI without tailored responses (JavaFX Part 3)
  Implement ChatBot Window layout (JavaFX Part 2)
  Fix bug NoClassDefFound Error
  Fix Checkstyle errors
  Fix coding standards from Git PRs
  Follow Coding Standard: categorise import statement by package
  Add JavaDoc Comments

# Conflicts:
#	src/main/java/seedu/duke/Duke.java
#	src/main/java/seedu/duke/task/Task.java

Commit:
Add Find feature

Commit:
Generate new JAR file

Commit:
Add Assertions to all classes

Exceptions and edge cases are enforced by try-catch blocks

Try-catch blocks are difficult to follow and contain many lines

Add assertions to make program thought-process clear, and demonstrate
awareness of cases where ArrayList.get(-1) or
NullPointerExceptions are denied

Let's,
* update all the classes to include Assertions

Commit:
Fix ambiguous variable names

Insufficient description in variable names

Proper variable naming helps with debugging and reading of code by
other programmers

Let's,
* rectify variable names such as response -> dukeResponse,
* splitAgain -> commandAndDescription
* number -> numberToDelete

Commit:
Add error handling for DateTimeParseException

starsia committed:
Merge pull request #2 from starsia/branch-A-Assertions

Add Assertions to all classes

Commit:
Merge branch 'master' of https://github.com/starsia/ip into branch-A-CodeQuality

* 'master' of https://github.com/starsia/ip:
  Add Assertions to all classes

# Conflicts:
#	src/main/java/seedu/duke/DialogBox.java
#	src/main/java/seedu/duke/Duke.java
#	src/main/java/seedu/duke/Storage.java

starsia committed:
Merge pull request #3 from starsia/branch-A-CodeQuality

Improve Code Quality

Commit:
Add shorter keywords for commands

Make the command syntax more flexible.

e.g., t can be shorter alias for todo.

Commit:
Add Ui.png

Commit:
Update `docs\READ.ME` with GFMD

Commit:
Change all Duke instances to Klara

Commit:
Add non-trivial JUnit tests to Task and TaskList classes

Commit:
Fix command shortcut to only accept single letters

Commit:
Fix dialog box size limit upon 'list' command

starsia committed:
Revise README for Duke project details

Updated project description and setup instructions.

starsia committed:
Change project title from 'Duke project template' to 'Duke'

Updated project title in README.
